bootstrap/kubeadm/api/v1alpha2/conversion.go

```diff
@@ -36,27 +36,34 @@ func (src *KubeadmConfig) ConvertTo(dstRaw conversion.Hub) error {
 		return err
 	}
 
+	if len(dst.Annotations) == 0 {
+		dst.Annotations = nil
+	}
```

Comment:
Why do we need to reset annotations if they are 0?

Reply:
Interesting, I wonder if something else is setting this / if the problem is elsewhere 🤔 

Wouldn't this have to happen in other types as well?

Reply:
Without looking super closely, I think it's happening because we're not using the fuzzer. I don't think the fuzzer will allow for pointer fields to be nil with a probabilistic chance (or at least that was my experience debugging the failure cases and trying to use the fuzzer here). So It will populate it with garbage which gets successfully round tripped because there's never a case with zero annotations populated.

I ended up not using the fuzzer because I didn't quickly find a good way to fuzz the union-y struct where only one value should be populated. The fuzzer always populates everything it seems so the conversion is technically correct but semantically wrong. I know go-fuzz and other tools do support probabilistic chances for pointers, not sure about the validation. Would have to look at the upstream k8s library and what it does under the hood.

Reply:
I can take another look at that if this smells fishy to you

Reply:
Could update the tests to run validation on fuzzed structs before attempting conversion, bit of a mixing of concerns but not a terrible idea

Reply:
I we remove these 3 lines from the conversion, and keep the test changes, does it pass?

Reply:
I think I sound confusing because I rewrote the whole function when I figured this out. Here's a minimal diff to see the problem. https://diffy.org/diff/mrrlhjkagf1j9p5c3yeukgldi

Before:
```go
ace@ace-vm:~/go/src/sigs.k8s.io/cluster-api$ go test -v ./bootstrap/kubeadm/api/v1alpha2/... -run TestConvertKubeadmConfig -count=1
=== RUN   TestConvertKubeadmConfig
=== RUN   TestConvertKubeadmConfig/from_hub
=== RUN   TestConvertKubeadmConfig/from_hub/preserves_fields_from_hub_version
--- PASS: TestConvertKubeadmConfig (0.00s)
    --- PASS: TestConvertKubeadmConfig/from_hub (0.00s)
        --- PASS: TestConvertKubeadmConfig/from_hub/preserves_fields_from_hub_version (0.00s)
PASS
ok  	sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha2	0.013s
```

After:
```go
ace@ace-vm:~/go/src/sigs.k8s.io/cluster-api$ go test -v ./bootstrap/kubeadm/api/v1alpha2/... -run TestConvertKubeadmConfig -count=1
=== RUN   TestConvertKubeadmConfig
=== RUN   TestConvertKubeadmConfig/from_hub
=== RUN   TestConvertKubeadmConfig/from_hub/preserves_fields_from_hub_version
--- FAIL: TestConvertKubeadmConfig (0.00s)
    --- FAIL: TestConvertKubeadmConfig/from_hub (0.00s)
        --- FAIL: TestConvertKubeadmConfig/from_hub/preserves_fields_from_hub_version (0.00s)
            conversion_test.go:81:   map[string]string(
                - 	nil,
                + 	{},
                  )
FAIL
FAIL	sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha2	0.013s
FAIL
```

I'll update to reflect.

Reply:
Other thoughts: annotations are omitempty, so these are equivalent on serialization

Reply:
reverted this based on serialization, IMO that's really what counts here. The semantics are correct.

Reply:
makes sense, thanks!